File: merge sort.cpp

#include <iostream>
using namespace std;



void merge(int array[], int const left, int const mid, int const right)
{
	auto const subAOne = mid - left + 1;
	auto const subATwo = right - mid;

	
	auto *leftArray = new int[subAOne],
		*rightArray = new int[subATwo];

	
	for (auto k = 0; k < subAOne; k++)
		leftArray[k] = array[left + k];
	for (auto f = 0; f < subATwo; f++)
		rightArray[f] = array[mid + 1 + f];

	auto indexOfSubAOne = 0, 
		indexOfSubATwo = 0; 
	int indexOfMergedA = left; 

	
	while (indexOfSubAOne < subAOne && indexOfSubATwo < subATwo) {
		if (leftArray[indexOfSubAOne] <= rightArray[indexOfSubATwo]) {
			array[indexOfMergedA] = leftArray[indexOfSubAOne];
			indexOfSubAOne++;
		}
		else {
			array[indexOfMergedA] = rightArray[indexOfSubATwo];
			indexOfSubATwo++;
		}
		indexOfMergedA++;
	}
	
	while (indexOfSubAOne < subAOne) {
		array[indexOfMergedA] = leftArray[indexOfSubAOne];
		indexOfSubAOne++;
		indexOfMergedA++;
	}
	while (indexOfSubATwo < subATwo) {
		array[indexOfMergedA] = rightArray[indexOfSubATwo];
		indexOfSubATwo++;
		indexOfMergedA++;
	}
}


void mergeSort(int array[], int const begin, int const end)
{
	if (begin >= end)
		return; 
	auto mid = begin + (end - begin) / 2;
	mergeSort(array, begin, mid);
	mergeSort(array, mid + 1, end);
	merge(array, begin, mid, end);
}


void printArray(int A[], int size)
{
	for (auto k = 0; k < size; k++)
		cout << A[k] << " ";
}


int main()
{
	int arr[] = { 23, 17, 2, 67, 10, 88 };
	auto arr_size = sizeof(arr) / sizeof(arr[0]);

	cout << "Given array is \n";
	printArray(arr, arr_size);

	mergeSort(arr, 0, arr_size - 1);

	cout << "\nSorted array is \n";
	printArray(arr, arr_size);
	return 0;
}

File: merge sort.py


def merge(arr, p, v, c):
	n1 = v- p + 1
	n2 = c - v

	
	P = [0] * (n1)
	C= [0] * (n2)

	
	for i in range(0, n1):
		P[i] = arr[p + i]

	for j in range(0, n2):
		C[j] = arr[v + 1 + j]

	
	i = 0	 
	j = 0	 
	k = p	 
	while i < n1 and j < n2:
		if P[i] <= C[j]:
			arr[k] = P[i]
			i += 1
		else:
			arr[k] = C[j]
			j += 1
		k += 1


	while i < n1:
		arr[k] = P[i]
		i += 1
		k += 1

	
	while j < n2:
		arr[k] = C[j]
		j += 1
		k += 1


def mergeSort(arr, p, r):
	if p < r:

		
		m = p+(r-p)//2

		
		mergeSort(arr, p, m)
		mergeSort(arr, m+1, r)
		merge(arr, p, m, r)



arr = [44, 19, 33, 2, 9, 43]
n = len(arr)
print("Given array is")
for i in range(n):
	print("%d" % arr[i],end=" ")

mergeSort(arr, 0, n-1)
print("\n\nSorted array is")
for i in range(n):
	print("%d" % arr[i],end=" ")


